pulp_container/app/registry_api.py

```diff
@@ -475,6 +479,161 @@ def get_queryset(self, *args, **kwargs):
         return (public_repositories | accessible_repositories).distinct()
 
 
+class FlatpakIndexDynamicView(APIView):
+    """
+    Handles requests to the /index/dynamic endpoint
+    """
+
+    def get(self, request):
+        unsatisfiable = False
+        req_repository = None
+        req_tags = set()
+        req_os = None
+        req_architecture = None
+        req_label_exists = {"org.flatpak.ref"}
+        req_label_values = {}
+        for key, values in request.query_params.lists():
+            if key == "repository":
+                req_repository = values[0]
+                if any(v != req_repository for v in values[1:]):
```

Comment:
From private communication between @owtaylor and me:
"sbergman:  At <https://github.com/owtaylor/flagstate/blob/master/docs/protocol.md#requests>, for an image to match the given query parameters means that it matches their conjunction, right?  (i.e., `?architecture=amd64&architecture=arm64` will necessarily return an empty list, not a list containing images that mach either of the architectures)
"Owen Taylor:  Yeah, I think so. I'm not sure I really considered repeated arguments."

(Yet something else to clarify in <https://github.com/owtaylor/flagstate/pull/6>.  Will do.)

So that's why I check in all these cases that if a query parameter like `repository` is given multiple times, that all the given values are the same, or else the request is unsatisfiable.

Reply:
gotcha, this info is helpful

Reply:
really depends on the client's expectation but what we could do instead of checking all the cases of repo query param to have same values, just pick the first value and ignore rest

Reply:
> From private communication between @owtaylor and me:
> "sbergman: At https://github.com/owtaylor/flagstate/blob/master/docs/protocol.md#requests, for an image to match the given query parameters means that it matches their conjunction, right? (i.e., `?architecture=amd64&architecture=arm64` will necessarily return an empty list, not a list containing images that mach either of the architectures)
> "Owen Taylor: Yeah, I think so. I'm not sure I really considered repeated arguments."

So, I dug through the Flagstate code, and the interpretation for repeated arguments is OR rather than AND. Which is definitely also the more useful interpretation.  My preference would be:

 * Document the OR implementation in the protoocl
 * If it's easy to implement the OR in the Pulp code, do that
 * Otherwise, error out (with a 501 Not Implemented perhaps) if a repeated term is found

The Flatpak client will *not* pass in repeated terms, so there is no need to go through a lot of work to implement them.

Reply:
So, just to clarify, the rule would be:  An image I matches a set S of query parameters iff:  For each K=V in S:  There exists a K'=V' in S with:  K==K' and I matches V'.

(That is, `?label:eat=apple&label:eat=orange&label:drink=water` matches all images that have both: (a) an `eat` label of either `apple` or `orange`; and (b) a `drink` label of `water`.)

Reply:
addressed with <https://github.com/pulp/pulp_container/pull/1301/commits/719a9f599ba647de371a21450ab45f8c2af52366> "Fix logic for repeated query parameters" now